crates/astria-composer/src/searcher/executor.rs

```diff
@@ -176,18 +227,75 @@ impl Executor {
-            if let Err(e) = self.sign_and_submit(bundle).await {
-                // FIXME: currently this will fail both when there is an issue with the nonce and
-                // when unable to reach the sequencer. As there is currently no error returned by
-                // the sequencer for invalid nonces, there is nothing to handle. This should be
-                // changed after #364 is merged in a followup PR to handle nonce failues.
-                error!(
-                    error.message = %e,
-                    error.cause_chain = ?e,
-                    "failed submitting the bundle to the sequencer; bailing",
-                );
-                break;
+        let mut submission_fut = Fuse::terminated();
+        let mut nonce_fut = Fuse::terminated();
+        loop {
+            select! {
+                // receive new bundle for processing
+                Some(bundle) = self.executor_rx.recv(), if nonce_fut.is_terminated() && submission_fut.is_terminated() => {
+                    debug!(bundle = ?bundle, "executor received bundle for processing");
+                    nonce_fut = self.attach_nonce_fut(bundle);
+                },
+                // attach nonce
+                ret = &mut nonce_fut, if !nonce_fut.is_terminated() => {
+                    match ret {
+                        Ok(unsigned_tx) => {
+                            // update self.nonce for next transaction
+                            *Arc::get_mut(&mut self.nonce).expect("should only be called once at a time") =
+                                Some(unsigned_tx.nonce + 1);
+
+                            submission_fut = self.sign_and_submit_fut(unsigned_tx);
+                        }
+                        Err(e) => {
+                            error!(error.msg = %e, "failed to retrieve nonce from sequencer; executor shutting down");
+                            break;
+                        }
+                    }
+                }
+                // submit to sequencer
+                ret = &mut submission_fut, if !submission_fut.is_terminated() => {
+                    match ret {
+                        Err(ExecutionError::InvalidNonce(transaction)) => {
+                            let nonce = transaction.unsigned_transaction().nonce;
+                            warn!(
+                                nonce,
+                                "invalid nonce error returned from sequencer; retrieving new nonce and \
+                                 resubmitting the transaction"
+                            );
+                            nonce_fut = self.bundle_resubmission_from_transaction(transaction);
+                        }
+                        Err(ExecutionError::UnknownDeliverTxFailure {
+                            code,
+                            transaction,
+                        }) => {
+                            warn!(
+                                code=?code,
```

Comment:
Please don't use debug formatting in events, it's very noisy. AFAIK `AbciCode` implements `Display`:
```suggestion
                                %code,
```

Reply:
this is a `tendermint::abci::Code` though, which does not implement Display. Not sure what the solution here is. Should this be cast into an `AbciCode` at some point?

Reply:
use `AbciCode::from_tendermint()`?

Reply:
not every `tendermint::abci::Code` can be mapped to an `AbciCode` though

Reply:
In fact, `AbciCode` did not implement `Display`. We are printing with `AbciCode::value` now.

Also, I renamed `AbciCode::from_tendermint` to `AbciCode::from_cometbft`.